### PR TITLE
refactor(headers): introduce safe abstractions

### PR DESCRIPTION
**Pull request by gakonst**
Context:
* We were using `StageDB`, which is a wrapper over a mutable transaction, i.e. not really a DB, and has footguns (e.g. derefing to a tx makes it less clear to understand what is going on - my bad here when I introdcued this). This results into long-lived write transactions which we should think of as long-Mutexes, obviously bad.

Solution:
* We refactor each Stage to instead take a `DB: Database` and instantiate safe wrappers over it which proceed to give us methods for each action we want. For Headers, there is for example `HeadersReader` and `HeadersWriter`. These abstractions are going to be consumed by anyone that wants to use the DB, whether it's a stage, or a DB client.

TODO:
1. Refactor these safe abstractions in `storage/reth-providers`

### CODE REVIEW

**Comment by onbjerg** [comment]
This is not entirely correct since `db.update(f)` opens up a transaction, calls `f` and then commits the transaction. Since headers, total difficulty and the stage progress are now written in 3 separate transactions, we can end up in an inconsistent state. For example, we might write 500 headers, crash and then restart the node. At this point, the checkpoint does not reflect that we wrote 500 headers, and the `HeaderTD` table is out of sync with the `Headers` table as well.

TL;DR Rule of thumb: When data is committed to the database, *all* state has to be 100% consistent at that point.

**Comment by gakonst** [comment]
Closing. We decided that the StageDB abstraction is correct, in managing the lifecycle of things as-is. 

We don't love:
1. The name `StageDB`: To convert to `Transaction<_, DB>`
2. `StageDB` has too many things: Write wrappers around the `Transaction` and implement them as providers.
3. `StageDB` only supports `TxMut` which means that we always take a write-lock when we use it.
  1. The write transaction should be instantiated as lazily as possible to reduce the "area" that the write-lock covers
  2. We should add support for read transactions inside of `Transaction` so that we don't open redundant write-locks when we don't need them.

Basically `StageDB` --> `Transaction` and its current semantics represent a `Mutex` where we'd like it to look more like a `RWLock`.
